python_modules/dagster/dagster/_core/definitions/asset_checks.py

```diff
@@ -0,0 +1,119 @@
+from typing import Any, Dict, Iterable, Iterator, Mapping, Set
+
+from dagster import _check as check
+from dagster._annotations import experimental, public
+from dagster._core.definitions.asset_check_spec import AssetCheckSpec
+from dagster._core.definitions.events import AssetKey
+from dagster._core.definitions.node_definition import NodeDefinition
+from dagster._core.definitions.resource_definition import ResourceDefinition
+from dagster._core.definitions.resource_requirement import (
+    ResourceAddable,
+    ResourceRequirement,
+    merge_resource_defs,
+)
+
+
+@experimental
+class AssetChecksDefinition(ResourceAddable):
+    """Defines a set of checks that are produced by the same op or op graph.
+
+    AssetChecksDefinition are typically not instantiated directly, but rather produced using a
+    decorator like :py:func:`@asset_check <asset>`.
+    """
+
+    _node_def: NodeDefinition
+    _resource_defs: Mapping[str, ResourceDefinition]
+    _specs_by_output_name: Mapping[str, AssetCheckSpec]
+
+    def __init__(
+        self,
+        node_def: NodeDefinition,
+        resource_defs: Mapping[str, ResourceDefinition],
+        asset_keys_by_input_name: Mapping[str, AssetKey],
+        specs_by_output_name: Mapping[str, AssetCheckSpec],
+        # if adding new fields, make sure to handle them in the get_attributes_dict method
+    ):
+        self._node_def = node_def
+        self._resource_defs = resource_defs
+        self._asset_keys_by_input_name = check.dict_param(
+            asset_keys_by_input_name, "asset_keys_by_input_name", key_type=str, value_type=AssetKey
+        )
+        self._specs_by_output_name = check.dict_param(
+            specs_by_output_name, "specs_by_output_name", key_type=str, value_type=AssetCheckSpec
+        )
+
+    @public
+    @property
+    def node_def(self) -> NodeDefinition:
+        """The op or op graph that can be executed to check the assets."""
+        return self._node_def
+
+    @public
+    @property
+    def name(self) -> str:
+        return self.spec.name
+
+    @public
+    @property
+    def description(self) -> str:
+        return self.spec.description
+
+    @public
+    @property
+    def asset_key(self) -> AssetKey:
+        return self.spec.asset_key
+
+    @public
+    @property
+    def spec(self) -> AssetCheckSpec:
```

Comment:
If this is frequently access string construction in the invariant message might become expensive

Reply:
great point. fixing.